### PR TITLE
Migrate server side to new kibana plugin platform

### PR DESCRIPTION
**Pull request by zhongnansu**
*Issue #, if available:*
#6 
*Description of changes:*

- update kibana version number to `8.0.0` as of the master branch of elastic kibana
- new plugin definition shim: remove old `index.js` in root and add `plugin.ts` and `index.ts` to `server` and `public` folder
  ```
  src/plugins
  kibana-reports
    kibana.json
    public
      index.ts
      plugin.ts
    server
      index.ts
      plugin.ts
  ```
- input validation: `Joi`  → use `@kbn-config-schema`
  - https://github.com/elastic/kibana/tree/master/packages/kbn-config-schema
- communicate with Elasticsearch:  `callWithRequest()` → `context.core.elasticsearch.legacy.client.callAsInternalUser()` 
  - https://github.com/elastic/kibana/blob/master/src/core/MIGRATION.md#use-scoped-services
- De-couple from hapi.js server and request objects
  - `hapi h.file()` → `response.custom()/response.ok()`
- restructure the `route` layout
- Pass logger to route handler via context
- Update to better error handling

**Attention**
- For development purpose, I temporarily disable loading the client side module by setting the `ui: false` in `kibana.json`
- Added simple 

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: public/plugin.ts [hunk 1]
```diff
@@ -0,0 +1,58 @@
+import { i18n } from '@kbn/i18n';
```
Comment (davidcui-amzn) [inline public/plugin.ts:1]: copyright notice missing

File: public/index.ts [hunk 2]
```diff
@@ -0,0 +1,13 @@
+import './index.scss';
```
Comment (davidcui-amzn) [inline public/index.ts:1]: copyright missing

File: public/types.ts [hunk 3]
```diff
@@ -0,0 +1,9 @@
+import { NavigationPublicPluginStart } from '../../../src/plugins/navigation/public';
```
Comment (davidcui-amzn) [inline public/types.ts:1]: copyright

File: common/index.ts [hunk 4]
```diff
@@ -0,0 +1,4 @@
+export const PLUGIN_ID = 'opendistroKibanaReports';
+export const PLUGIN_NAME = 'opendistro_kibana_reports';
+
+export const API_PREFIX = '/api/v1/reporting';
```
Comment (seraphjiang) [inline common/index.ts:4]: add v1 to api endpoint is tricky, would suggest just use /api/reporting

File: package.json [hunk 5]
```diff
@@ -1,11 +1,11 @@
 {
-  "name": "opendistro-kibana-reports",
+  "name": "opendistro_kibana_reports",
   "version": "0.0.1",
   "description": "Kibana Reporting Plugin",
   "license": "Apache-2.0",
-  "main": "index.js",
+  "main": "target/plugins/opendistro_kibana_reports",
```
Comment (seraphjiang) [inline package.json:6]: not sure about this part.

File: public/plugin.ts [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+          depsStart as AppPluginStartDependencies,
+          params
+        );
+      },
+    });
+
+    // Return methods that should be available to other plugins
+    return {
+      getGreeting() {
+        return i18n.translate('examplePlugin.greetingText', {
```
Comment (seraphjiang) [inline public/plugin.ts:43]: remove example code

Comment (zhongnansu) [reply public/plugin.ts:43]: will remove every changes under /public, to keep this PR pure for server side migration.

File: public/types.ts (hunk 3 above)
Comment (zhongnansu) [reply public/types.ts:1]: will remove every changes under /public, to keep this PR pure for server side migration.

File: common/index.ts (hunk 4 above)
Comment (zhongnansu) [reply common/index.ts]: make sense

File: common/index.ts [hunk 7]
```diff
@@ -0,0 +1,4 @@
+export const PLUGIN_ID = 'opendistroKibanaReports';
+export const PLUGIN_NAME = 'opendistro_kibana_reports';
```
Comment (davidcui-amzn) [inline common/index.ts:2]: any reason its changed from `opendistro-kibana-reports`?

Comment (zhongnansu) [reply common/index.ts:2]: Not any big reasons, just that I saw we use `_` more often than `-` in the codebase, so I think it's better we keep the same style

File: package.json (hunk 5 above)
Comment (zhongnansu) [reply package.json]: You are right, changed it to `index.ts`

Comment (zhongnansu) [reply package.json]: Post the plugin install issue here https://github.com/opendistro-for-elasticsearch/kibana-reports/issues/40

**Pull request merged**